windows/src/developer/TIKE/main/UframeBitmapEditor.pas

```diff
@@ -817,29 +869,85 @@ procedure TframeBitmapEditor.LoadFromStream(Stream: TStream; IsIcon: Boolean);
   end;
   Modified := False;
   FEdit.IsTemp := False;
+
+  panReadOnlyIcons.Visible := not FIconCanBeEdited;
   panEdit.RePaint;
   panPreview.RePaint;
   SaveUndoBitmap;
 end;
 
-procedure TframeBitmapEditor.SaveToFile(FileName: WideString);
+(**
+  Loads a view of the icons available in a .ico file
+
+  Some of this code mirrors Vcl.Graphics ReadIcon
+  because WinAPI doesn't have an API to read .ico
+  files that supports multiple images in a single file;
+  while there are possible COM APIs available, we don't
+  want to add a dependency just to read the icon data.
+*)
+procedure TframeBitmapEditor.LoadReadOnlyIcon(Stream: TStream);
+type
+  PIconRecArray = ^TIconRecArray;
+  TIconRecArray = array[0..300] of TIconRec;
 var
-  Stream: TStream;
+  ci: TCursorOrIcon;
+  List: PIconRecArray;
+  HeaderLen: Integer;
+  i, w, h: Integer;
+  cd: string;
+  p: Int64;
 begin
+  memoReadOnlyIcons.Clear;
+  p := Stream.Position;
+  Stream.ReadBuffer(ci, SizeOf(TCursorOrIcon));
+  HeaderLen := SizeOf(TIconRec) * ci.Count;
+  List := AllocMem(HeaderLen);
   try
-    Stream := TFileStream.Create(FileName, fmCreate);
-  except
-    on E:EFileStreamError do
+    Stream.Read(List^, HeaderLen);
+    SetLength(FReadOnlyIcons, ci.Count);
+    for i := 0 to ci.Count - 1 do
     begin
-      ShowMessage(E.Message);
-      Exit;
+      Stream.Position := p;
+      FReadOnlyIcons[i] := TIcon.Create;
+      if List[i].Colors = 0
+        then cd := '24 bit'
+        else cd := IntToStr(List[i].Colors);
+      if List[i].Width = 0 then w := 256 else w := List[i].Width;
+      if List[i].Height = 0 then h := 256 else h := List[i].Height;
+
+      FReadOnlyIcons[i].SetSize(w, h);
+      FReadOnlyIcons[i].LoadFromStream(Stream);
```

Comment:
I take it that `FReadOnlyIcons` is a class field, and we're simply ensuring it's initialized properly upon a load of a project file?  (Noting this method's role in [`LoadFromStream`](https://github.com/keymanapp/keyman/pull/498/files#diff-1c98b4ad1b2560b643df7b494eb6967dR839) above.)

Reply:
Correct; see line 123; being a class member it is set to size=0 at object creation time so only if there are multiple icons in the file will this ever have more than zero members (it's a dynamic array of objects, not an object itself, so doesn't need construction or destruction).